File: specs/SPEC-9c11b38c/contracts/public-api.cs
/**
 * Unity Push Notification Plugin - Public API Specification
 *
 * 機能ID: SPEC-9c11b38c
 * 日付: 2025-10-30
 *
 * このファイルは公開API仕様を定義します（実装コードではありません）
 */

using System;
using UnityEngine;

namespace PushNotification
{
    /// <summary>
    /// Push Notification Plugin メインエントリポイント
    ///
    /// シングルトンとして提供され、SDK初期化・設定を管理します。
    ///
    /// 使用例:
    /// <code>
    /// var config = new PushNotificationConfig
    /// {
    ///     apiKey = "your-api-key",
    ///     apiEndpoint = "https://api.example.com"
    /// };
    /// PushNotificationManager.Initialize(config);
    /// </code>
    /// </summary>
    public static class PushNotificationManager
    {
        /// <summary>
        /// SDKを初期化します。
        ///
        /// ゲーム起動時に1回だけ呼び出してください。
        /// この関数は非同期で実行され、デバイストークン取得とREST API登録を行います。
        ///
        /// <param name="config">SDK設定オブジェクト</param>
        /// <exception cref="InvalidOperationException">すでに初期化済みの場合</exception>
        /// <exception cref="ArgumentException">不正な設定値の場合</exception>
        ///
        /// 使用例:
        /// <code>
        /// public class GameManager : MonoBehaviour
        /// {
        ///     void Start()
        ///     {
        ///         var config = new PushNotificationConfig
        ///         {
        ///             apiKey = "your-api-key",
        ///             apiEndpoint = "https://api.example.com",
        ///             enableLogging = true
        ///         };
        ///         PushNotificationManager.Initialize(config);
        ///     }
        /// }
        /// </code>
        /// </summary>
        public static void Initialize(PushNotificationConfig config);

        /// <summary>
        /// 通知ハンドラーを設定します。
        ///
        /// 通知受信・タップ時のカスタム処理を実装できます。
        /// 設定しない場合、デフォルトの通知表示のみ行われます。
        ///
        /// <param name="onReceived">通知受信時のコールバック（フォアグラウンド/バックグラウンド両方）</param>
        /// <param name="onOpened">通知タップ時のコールバック</param>
        /// <exception cref="InvalidOperationException">SDK未初期化の場合</exception>
        ///
        /// 使用例:
        /// <code>
        /// PushNotificationManager.SetNotificationHandler(
        ///     onReceived: (data) =>
        ///     {
        ///         Debug.Log($"Received: {data.title}");
        ///     },
        ///     onOpened: (data) =>
        ///     {
        ///         // カスタムデータから画面IDを取得して遷移
        ///         if (data.customData.TryGetValue("screenId", out string screenId))
        ///         {
        ///             NavigateToScreen(screenId);
        ///         }
        ///     }
        /// );
        /// </code>
        /// </summary>
        public static void SetNotificationHandler(
            Action<NotificationData> onReceived,
            Action<NotificationData> onOpened);

        /// <summary>
        /// プレイヤーアカウントIDを更新します。
        ///
        /// ユーザーログイン後にプレイヤーアカウントIDをデバイストークンに紐付けます。
        /// REST APIにPUTリクエストを送信し、サーバー側でトークン更新を行います。
        ///
        /// <param name="playerId">プレイヤーアカウントID</param>
        /// <exception cref="InvalidOperationException">SDK未初期化の場合</exception>
        /// <exception cref="ArgumentException">playerIdが空の場合</exception>
        ///
        /// 使用例:
        /// <code>
        /// // ユーザーログイン成功後
        /// PushNotificationManager.UpdatePlayerAccountId("player-12345");
        /// </code>
        /// </summary>
        public static void UpdatePlayerAccountId(string playerId);

        /// <summary>
        /// デバイストークンを登録解除します。
        ///
        /// ユーザーログアウト時やPush通知オプトアウト時に呼び出します。
        /// REST APIにDELETEリクエストを送信し、サーバー側でトークンを無効化します。
        ///
        /// <exception cref="InvalidOperationException">SDK未初期化の場合</exception>
        ///
        /// 使用例:
        /// <code>
        /// // ユーザーログアウト時
        /// PushNotificationManager.UnregisterToken();
        /// </code>
        /// </summary>
        public static void UnregisterToken();

        /// <summary>
        /// SDK初期化済みフラグを返します。
        /// </summary>
        /// <returns>初期化済みの場合true</returns>
        public static bool IsInitialized { get; }

        /// <summary>
        /// 現在のデバイストークンを取得します。
        /// </summary>
        /// <returns>デバイストークン（未取得の場合null）</returns>
        public static string CurrentDeviceToken { get; }

        /// <summary>
        /// 現在のプラットフォームを取得します。
        /// </summary>
        /// <returns>プラットフォーム種別</returns>
        public static PlatformType CurrentPlatform { get; }

        /// <summary>
        /// エラーコールバックを設定します（オプション）。
        ///
        /// SDK内部エラー発生時に通知を受け取ります。
        ///
        /// <param name="onError">エラーコールバック</param>
        ///
        /// 使用例:
        /// <code>
        /// PushNotificationManager.SetErrorHandler((error) =>
        /// {
        ///     Debug.LogError($"[Push SDK] {error.code}: {error.message}");
        /// });
        /// </code>
        /// </summary>
        public static void SetErrorHandler(Action<SDKError> onError);
    }

    /// <summary>
    /// プラットフォームブリッジ抽象インターフェース
    ///
    /// iOS/Android/Editorの各実装がこのインターフェースを実装します。
    /// 内部使用のみ、公開APIではありません。
    /// </summary>
    internal interface IPlatformBridge
    {
        /// <summary>
        /// Native SDKを初期化します。
        /// </summary>
        /// <param name="apiKey">APIキー</param>
        /// <param name="endpoint">APIエンドポイント</param>
        void Initialize(string apiKey, string endpoint);

        /// <summary>
        /// デバイストークンを取得します。
        /// </summary>
        /// <returns>デバイストークン（未取得の場合null）</returns>
        string GetDeviceToken();

        /// <summary>
        /// デバイストークンをREST APIに登録します。
        /// </summary>
        /// <param name="token">デバイストークン</param>
        void RegisterToken(string token);

        /// <summary>
        /// 通知コールバックを設定します。
        /// </summary>
        /// <param name="callback">JSON文字列を受け取るコールバック</param>
        void SetNotificationCallback(Action<string> callback);

        /// <summary>
        /// プレイヤーアカウントIDを更新します。
        /// </summary>
        /// <param name="playerId">プレイヤーアカウントID</param>
        void UpdatePlayerAccountId(string playerId);

        /// <summary>
        /// デバイストークンを登録解除します。
        /// </summary>
        void UnregisterToken();

        /// <summary>
        /// プラットフォーム種別を取得します。
        /// </summary>
        /// <returns>プラットフォーム種別</returns>
        PlatformType GetPlatformType();
    }

    /// <summary>
    /// Unity Editor専用 - 疑似通知送信ユーティリティ
    ///
    /// Unity Editorでのみ使用可能。実機ビルドでは無視されます。
    /// </summary>
#if UNITY_EDITOR
    public static class EditorNotificationSimulator
    {
        /// <summary>
        /// 疑似通知を送信します。
        ///
        /// Unity Editorでのテスト用に、疑似的に通知を発火させます。
        ///
        /// <param name="data">通知データ</param>
        ///
        /// 使用例:
        /// <code>
        /// var testData = new NotificationData
        /// {
        ///     title = "テスト通知",
        ///     body = "これはテストです",
        ///     customData = new Dictionary<string, string>
        ///     {
        ///         { "screenId", "event" }
        ///     }
        /// };
        /// EditorNotificationSimulator.SendNotification(testData);
        /// </code>
        /// </summary>
        public static void SendNotification(NotificationData data);

        /// <summary>
        /// 疑似デバイストークンを生成します。
        /// </summary>
        /// <returns>疑似トークン（UUID）</returns>
        public static string GenerateMockToken();
    }
#endif

    /// <summary>
    /// プラットフォームファクトリー
    ///
    /// 実行中のプラットフォームに応じて適切なIPlatformBridge実装を返します。
    /// 内部使用のみ。
    /// </summary>
    internal static class PlatformBridgeFactory
    {
        /// <summary>
        /// 現在のプラットフォームに対応するブリッジを作成します。
        /// </summary>
        /// <returns>プラットフォームブリッジ実装</returns>
        /// <exception cref="PlatformNotSupportedException">未対応プラットフォームの場合</exception>
        public static IPlatformBridge CreateBridge();
    }

    /// <summary>
    /// メインスレッドディスパッチャー
    ///
    /// Native SDKコールバック（別スレッド）をUnityメインスレッドで実行します。
    /// 内部使用のみ。
    /// </summary>
    internal class MainThreadDispatcher : MonoBehaviour
    {
        /// <summary>
        /// シングルトンインスタンスを初期化します。
        /// </summary>
        public static void Initialize();

        /// <summary>
        /// アクションをメインスレッドキューに追加します。
        /// </summary>
        /// <param name="action">実行するアクション</param>
        public static void Enqueue(Action action);

        /// <summary>
        /// メインスレッドで即座にアクションを実行します。
        /// </summary>
        /// <param name="action">実行するアクション</param>
        public static void Execute(Action action);
    }

    /// <summary>
    /// トークンストレージ
    ///
    /// デバイストークン・プレイヤーアカウントIDをPlayerPrefsで永続化します。
    /// 内部使用のみ。
    /// </summary>
    internal static class TokenStorage
    {
        /// <summary>
        /// トークン状態を保存します。
        /// </summary>
        /// <param name="state">トークン状態</param>
        public static void Save(TokenState state);

        /// <summary>
        /// トークン状態を読み込みます。
        /// </summary>
        /// <returns>トークン状態（存在しない場合null）</returns>
        public static TokenState Load();

        /// <summary>
        /// トークン状態をクリアします。
        /// </summary>
        public static void Clear();
    }

    /// <summary>
    /// ログユーティリティ
    ///
    /// SDK内部ログを管理します。enableLogging=falseの場合は出力しません。
    /// 内部使用のみ。
    /// </summary>
    internal static class Logger
    {
        /// <summary>
        /// ログ出力有効化フラグを設定します。
        /// </summary>
        /// <param name="enabled">有効化フラグ</param>
        public static void SetEnabled(bool enabled);

        /// <summary>
        /// デバッグログを出力します。
        /// </summary>
        /// <param name="message">メッセージ</param>
        public static void Debug(string message);

        /// <summary>
        /// 情報ログを出力します。
        /// </summary>
        /// <param name="message">メッセージ</param>
        public static void Info(string message);

        /// <summary>
        /// 警告ログを出力します。
        /// </summary>
        /// <param name="message">メッセージ</param>
        public static void Warning(string message);

        /// <summary>
        /// エラーログを出力します。
        /// </summary>
        /// <param name="message">メッセージ</param>
        public static void Error(string message);

        /// <summary>
        /// 例外ログを出力します。
        /// </summary>
        /// <param name="exception">例外</param>
        public static void Exception(Exception exception);
    }
}

File: specs/SPEC-628d6000/contracts/public-api.kt
/**
 * Android Push Notification SDK - Public API Specification
 *
 * 機能ID: SPEC-628d6000
 * 日付: 2025-10-30
 *
 * このファイルは公開API仕様を定義します（実装コードではありません）
 */

package com.example.pushnotification

import android.content.Context
import com.example.pushnotification.config.SDKConfig
import com.example.pushnotification.models.NotificationData

/**
 * Push Notification SDK メインエントリポイント
 *
 * シングルトンオブジェクトとして提供され、SDK初期化・設定を管理します。
 *
 * ## 使用例:
 * ```kotlin
 * val config = SDKConfig(
 *     apiKey = "your-api-key",
 *     apiEndpoint = "https://api.example.com"
 * )
 * PushNotificationSDK.initialize(context, config)
 * ```
 */
object PushNotificationSDK {

    /**
     * SDKを初期化します。
     *
     * アプリケーション起動時（Application.onCreate()）に1回だけ呼び出してください。
     * この関数は非同期で実行され、FCMトークン取得とREST API登録を行います。
     *
     * @param context Applicationコンテキスト
     * @param config SDK設定オブジェクト
     * @throws IllegalStateException すでに初期化済みの場合
     * @throws IllegalArgumentException 不正な設定値の場合
     *
     * ## 使用例:
     * ```kotlin
     * class MyApplication : Application() {
     *     override fun onCreate() {
     *         super.onCreate()
     *         val config = SDKConfig(
     *             apiKey = "your-api-key",
     *             apiEndpoint = "https://api.example.com",
     *             enableLogging = true
     *         )
     *         PushNotificationSDK.initialize(this, config)
     *     }
     * }
     * ```
     */
    fun initialize(context: Context, config: SDKConfig)

    /**
     * プレイヤーアカウントIDを更新します。
     *
     * ユーザーログイン後にプレイヤーアカウントIDをデバイストークンに紐付けます。
     * REST APIにPUTリクエストを送信し、サーバー側でトークン更新を行います。
     *
     * @param playerId プレイヤーアカウントID
     * @throws IllegalStateException SDK未初期化の場合
     * @throws IllegalArgumentException playerIdが空の場合
     *
     * ## 使用例:
     * ```kotlin
     * // ユーザーログイン成功後
     * PushNotificationSDK.updatePlayerAccountId("player-12345")
     * ```
     */
    fun updatePlayerAccountId(playerId: String)

    /**
     * デバイストークンを登録解除します。
     *
     * ユーザーログアウト時やPush通知オプトアウト時に呼び出します。
     * REST APIにDELETEリクエストを送信し、サーバー側でトークンを無効化します。
     *
     * @throws IllegalStateException SDK未初期化の場合
     *
     * ## 使用例:
     * ```kotlin
     * // ユーザーログアウト時
     * PushNotificationSDK.unregisterToken()
     * ```
     */
    fun unregisterToken()

    /**
     * 通知ハンドラーを設定します。
     *
     * 通知受信・タップ時のカスタム処理を実装できます。
     * 設定しない場合、デフォルトの通知表示のみ行われます。
     *
     * @param handler NotificationHandlerインターフェース実装
     * @throws IllegalStateException SDK未初期化の場合
     *
     * ## 使用例:
     * ```kotlin
     * PushNotificationSDK.setNotificationHandler(object : NotificationHandler {
     *     override fun onNotificationReceived(data: NotificationData) {
     *         Log.d("Push", "Received: ${data.title}")
     *     }
     *
     *     override fun onNotificationOpened(data: NotificationData) {
     *         // カスタムデータから画面IDを取得して遷移
     *         val screenId = data.customData?.get("screenId")
     *         navigateToScreen(screenId)
     *     }
     * })
     * ```
     */
    fun setNotificationHandler(handler: NotificationHandler)

    /**
     * SDK初期化済みフラグを返します。
     *
     * @return 初期化済みの場合true
     */
    fun isInitialized(): Boolean

    /**
     * 現在のデバイストークンを取得します。
     *
     * @return デバイストークン（未取得の場合null）
     */
    fun getCurrentToken(): String?
}

/**
 * 通知イベントハンドラーインターフェース
 *
 * 通知受信・タップ時のカスタム処理を実装します。
 */
interface NotificationHandler {

    /**
     * 通知受信時に呼び出されます。
     *
     * フォアグラウンド/バックグラウンド両方で呼び出されます。
     * この関数内で重い処理を行うと通知表示が遅延するため、
     * 必要に応じて別スレッドで実行してください。
     *
     * @param data 通知データ
     */
    fun onNotificationReceived(data: NotificationData)

    /**
     * 通知タップ時に呼び出されます。
     *
     * ディープリンク処理や画面遷移をこの関数内で実装します。
     *
     * @param data 通知データ
     */
    fun onNotificationOpened(data: NotificationData)

    /**
     * トークン登録成功時に呼び出されます（オプション）。
     *
     * デフォルト実装は何もしません。
     *
     * @param token 登録されたデバイストークン
     */
    fun onTokenRegistered(token: String) {
        // デフォルト実装: 何もしない
    }

    /**
     * トークン登録失敗時に呼び出されます（オプション）。
     *
     * デフォルト実装は何もしません。
     *
     * @param error エラー情報
     */
    fun onTokenRegistrationFailed(error: SDKError) {
        // デフォルト実装: 何もしない
    }
}

/**
 * 通知チャンネル管理API
 *
 * Android 8.0以降の通知チャンネルを管理します。
 */
object ChannelManager {

    /**
     * 通知チャンネルを作成します。
     *
     * @param id チャンネルID（ユニーク）
     * @param name チャンネル名（ユーザーに表示される）
     * @param importance 通知重要度（IMPORTANCE_HIGH, IMPORTANCE_DEFAULT, IMPORTANCE_LOW）
     * @throws IllegalArgumentException 不正なパラメータの場合
     *
     * ## 使用例:
     * ```kotlin
     * ChannelManager.createChannel(
     *     id = "event_notifications",
     *     name = "イベント通知",
     *     importance = NotificationManager.IMPORTANCE_HIGH
     * )
     * ```
     */
    fun createChannel(id: String, name: String, importance: Int)

    /**
     * 通知チャンネルを削除します。
     *
     * @param id チャンネルID
     */
    fun deleteChannel(id: String)

    /**
     * すべての通知チャンネルを取得します。
     *
     * @return チャンネルIDとチャンネル名のマップ
     */
    fun getAllChannels(): Map<String, String>
}

/**
 * Unity/Unreal Engine JNIブリッジ用API
 *
 * Unity C#やUnreal Engine C++から呼び出すための静的関数を提供します。
 */
object PushNotificationJNI {

    /**
     * SDKを初期化します（JNI用）。
     *
     * @param context Androidコンテキスト
     * @param apiKey APIキー
     * @param apiEndpoint APIエンドポイント
     * @param enableLogging ロギング有効化フラグ
     *
     * ## Unity C#から呼び出す例:
     * ```csharp
     * AndroidJavaClass javaClass = new AndroidJavaClass(
     *     "com.example.pushnotification.PushNotificationJNI"
     * );
     * AndroidJavaObject activity = new AndroidJavaClass("com.unity3d.player.UnityPlayer")
     *     .GetStatic<AndroidJavaObject>("currentActivity");
     *
     * javaClass.CallStatic("initialize", activity, "api-key", "https://api.example.com", true);
     * ```
     */
    @JvmStatic
    fun initialize(context: Context, apiKey: String, apiEndpoint: String, enableLogging: Boolean)

    /**
     * プレイヤーアカウントIDを更新します（JNI用）。
     *
     * @param playerId プレイヤーアカウントID
     */
    @JvmStatic
    fun updatePlayerAccountId(playerId: String)

    /**
     * デバイストークンを登録解除します（JNI用）。
     */
    @JvmStatic
    fun unregisterToken()

    /**
     * 通知コールバックを設定します（JNI用）。
     *
     * @param callback JSON文字列を受け取るコールバック関数
     *
     * ## Unity C#から呼び出す例:
     * ```csharp
     * javaClass.CallStatic("setNotificationCallback", new Action<string>((json) => {
     *     Debug.Log("Notification received: " + json);
     *     // JSONをパースして処理
     * }));
     * ```
     */
    @JvmStatic
    fun setNotificationCallback(callback: (String) -> Unit)

    /**
     * 現在のデバイストークンを取得します（JNI用）。
     *
     * @return デバイストークン（未取得の場合null）
     */
    @JvmStatic
    fun getCurrentToken(): String?
}

/**
 * SDKエラー情報
 */
data class SDKError(
    val code: ErrorCode,
    val message: String,
    val cause: Throwable? = null
) {
    enum class ErrorCode {
        INITIALIZATION_FAILED,
        TOKEN_FETCH_FAILED,
        TOKEN_REGISTRATION_FAILED,
        TOKEN_UPDATE_FAILED,
        TOKEN_DELETE_FAILED,
        NOTIFICATION_DISPLAY_FAILED,
        IMAGE_LOAD_FAILED,
        API_NETWORK_ERROR,
        API_AUTHENTICATION_ERROR,
        API_SERVER_ERROR,
        API_TIMEOUT_ERROR,
        UNKNOWN_ERROR
    }
}
